src/persist-client/src/impl/machine.rs

```diff
@@ -897,16 +908,27 @@ mod tests {
             .expect("scan failed");
         // Make sure we constructed the key correctly.
         assert!(consensus_entries.len() > 0);
-        // Make sure the number of entries is bounded.
-        //
-        // In practice, this is always 1 right now, but when we implement
-        // incremental state, it will be something like log(NUM_BATCHES).
-        let max_entries = usize::cast_from(NUM_BATCHES.next_power_of_two().trailing_zeros());
-        assert!(
-            consensus_entries.len() <= max_entries,
-            "expected at most {} entries got {}",
-            max_entries,
-            consensus_entries.len()
-        );
+        // Make sure the number of entries is bounded. (I think we could work
+        // out a tighter bound than this, but the point is only that it's
+        // bounded).
+        let max_entries = 2 * usize::cast_from(NUM_BATCHES.next_power_of_two().trailing_zeros());
+
+        // The truncation is done in the background, so wait until the assertion
+        // passes, giving up and failing the test if that takes "too much" time.
```

Comment:
I don't think it's worth it. having a more complicated protocol so we don't have to loop until an assertion passes seems like the wrong set of tradeoffs to me. personally, I'd rather move the bit that sends off the GarbageCollect call into the WriteHandle and ReadHandle (see how compaction works now), which would make it possible to be deterministic here (plus allow coverage as a datadriven test). but for some really dumb reasons, we'd have to plumb the gc requests back from basically every method on machine (compaction gets away with it because the only thing that can result in a compaction is compare_and_append).

I'd love to have something better here myself, but I beat my head on this for an hour or two and didn't come up with anything better than what you see here :)

Reply:
Fair! I'm struggling to come up with something as well, carry on